### ISSUE TITLE
JetStream Simplified API continuous polling with `Consume()` and `Messages()`

### ISSUE THREAD

**Issue by mdawar** [nats-io/nats.go#1338]
First thank you for the new JetStream API, it's very easy to use and really provides a better DX.

I have a few questions about continuous polling with `Consume()` and `Messages()` because I had different experiences with both even though the docs say that they're interchangeable:

> Both Messages() and Consume() methods perform exactly the same optimizations and can be used interchangeably.

I'm using NATS for a job queue with workers running on multiple servers and each worker can be set to process multiple jobs concurrently that might take between a couple of seconds to a couple of minutes to finish.

This is how I'm creating the stream and the consumer:

```go
stream, _ := js.CreateStream(ctx, jetstream.StreamConfig{
	Name:      "jobs",
	Subjects:  []string{"jobs.>"},
	Retention: jetstream.WorkQueuePolicy,
})

consumer, _ := stream.CreateOrUpdateConsumer(ctx, jetstream.ConsumerConfig{
	Durable:       "JobsConsumer",
	ReplayPolicy:  jetstream.ReplayInstantPolicy,
	AckPolicy:     jetstream.AckExplicitPolicy,
	AckWait:       5 * time.Minute,
	MaxAckPending: -1,
})
```

With `Consume()`, I have used the code from the [`js-parallel-consume`](https://github.com/nats-io/nats.go/blob/97ccbd884fa75a3a5e86bef0ee4c9374dc0dd2b5/examples/jetstream/js-parallel-consume/main.go#L59) example:
```go
for i := 0; i < 5; i++ {
	cc, err := cons.Consume(
		func(consumeID int) jetstream.MessageHandler {
			return func(msg jetstream.Msg) {
				fmt.Printf("Received msg on consume %d\n", consumeID)
                                time.Sleep(2 * time.Second) // Simulate work (Using 30s for slow workers)
				msg.Ack()
			}
		}(i),
                jetstream.PullMaxMessages(1),
		jetstream.ConsumeErrHandler(func(consumeCtx jetstream.ConsumeContext, err error) {
			fmt.Printf("consume error: %v\n", err)
		}),
	)

	if err != nil {
		log.Fatal(err)
	}

	defer cc.Stop()
}
```

I'm running locally 4 workers each running `Consume()` 5 times as shown in the loop above, 2 of these workers are set to simulate work for 30 seconds each and 2 others for 2 seconds only.

I'm publishing 100 messages for testing this setup:

```go
for i := 0; i < 100; i++ {
	time.Sleep(50 * time.Millisecond)
	fmt.Println("Publishing job", i)

	_, err := js.Publish(ctx, "jobs.info", []byte(fmt.Sprintf("Job payload: %d", i)))

	if err != nil {
		log.Printf("cannot publish job #%d: %v", i, err)
	}
}
```

My observations:
- The workers consume messages in order of their connection (I don't know if the ID is being used for distribution)
- All the `Consume` callbacks are also run in order on each worker before other workers start consuming messages
- I get `nats: no heartbeat received` errors on the slow workers with `Consume()` with sleep for 30s after the callbacks return
- After these errors the slow workers seem to be pulling messages more than `PullMaxMessages` which is set to `1`, I was expecting these messages to be processed by other idle workers (I don't know if this is a misconfiguration on my side).

I have solved all of these issues by using `Messages()`:

```go
iter, err := consumer.Messages(
	jetstream.PullMaxMessages(1),
	jetstream.WithMessagesErrOnMissingHeartbeat(true), // Added for consistency with Consume()
)
if err != nil {
	log.Fatal(err)
}

defer iter.Stop()

parallelConsumers := 5
sem := make(chan struct{}, parallelConsumers)

for {
	sem <- struct{}{}

	go func() {
		defer func() {
			<-sem
		}()

		msg, err := iter.Next()
		if err != nil {
			fmt.Println("next err: ", err)
			return
		}

		fmt.Printf("Received msg: %s\n", msg.Data())
		time.Sleep(2 * time.Second) // Simulate work
		msg.Ack()
	}()
}
```

Now the workers consume the messages equally, and without heartbeat errors.
And even if I'm running 5 consumers (calling `Next()`) on each worker, the messages are pulled by the available worker goroutines fairly.

So it seems that `Consume()` and `Messages()` do not have the same performance, unless I have been doing something wrong on my side.

My question are:
- Are the `Consume()` callbacks supposed to be long running or not? Heartbeats are missed when they're long running.
- How messages are distributed when using pull consumers? It seems the consumer ID or the order of connection affects this.
- In my `Messages()` example which is worker as intended, is it OK to call `Next` for multiple goroutines? I'm assuming yes because of the mutex used in the code.

Used versions:
- Go: `v1.20.5` (linux/amd64)
- NATS server: `v2.9.19`
- nats.go: `v1.27.1`

Demo repository:
https://github.com/mdawar/jetstream-api-demos/tree/main/workqueue

Thank you.

**Comment by piotrpio** [comment]
Hey @mdawar, thanks for the issue - this to me looks like a bug, not misconfiguration, thank you so much for the detailed explanation and code samples, that's extremely helpful. I'm looking at this and will let you know as soon as I have some conclusions.

**Comment by piotrpio** [comment]
Okay, so I looked at you demo repository (nicely done!) and I believe I have some answers:

Firstly, the comment in `README.md` about `Consume()` and `Messages()` being 100% interchangeable is not precise, sorry about that. While for the most part it is true that the same optimizations (pre-buffering, sending concurrent pull requests etc.), there is one big difference, which becomes especially apparent when you start using those methods for work queues. Because both of those methods work on pull consumers, both of them have to send pull requests - we are sending new pull requests on certain thresholds (by default it's less than 50% of `MaxMessages`). Now the difference between `Consume()` and `Messages()` is **when** we are checking the threshold and sending new pull requests

- In `Consume()`, because you're supplying a callback, we need to determine whether a new pull request should be sent right after handling the message. Because of that, e.g. when you're using `PullMaxMessages(1)`, a new pull request is triggered as soon as the previous message is processed - this pull request will patiently wait for a new message from server when it eventually arrives.
- In `Messages()`, we do things a bit differently - we are checking the threshold and sending a new pull request (if needed) a the beginning of `Next()` method. That gives a bit more control over when new pull requests are sent to the server, which may be what's more convenient for work queue scenarios (server only sends new message after you call `Next()`)

So, to answer your first question - I will include those differences in readme, but in the meantime for work queues I would suggest using `Messages()`.

Moving on:
> The workers consume messages in order of their connection (I don't know if the ID is being used for distribution)
> All the Consume callbacks are also run in order on each worker before other workers start consuming messages

From what I saw, this is a byproduct of what I explained above. The server distributes messages across pull requests in round-robin, in the order in which they were sent. `Consume()` sends those pull requests right after processing the previous message (and the initial one as soon as you invoke `Consume()`), so naturally the will be appearing in order. In `Messages()`, because new pull request is only sent when you call `Next()`, I would assume the calls to the server may come in a slightly different order. Try adding any jitter to your processing time and you should notice that messages are fairly distributed.

> I get nats: no heartbeat received errors on the slow workers with Consume() with sleep for 30s after the callbacks return

That is a bug, thanks for discovering it. I will create a PR with fix soon. This problem occurs when handling a single messages takes more than 2 * `IdleHeartbeat` (which is 5 seconds by default). I will let you know when it's fixed.

> After these errors the slow workers seem to be pulling messages more than PullMaxMessages which is set to 1, I was expecting these messages to be processed by other idle workers (I don't know if this is a misconfiguration on my side).

That one is related to the issue above. This is expected behavior in a way. We are re-sending pull requests if there is a "no heartbeat" error, so that we are not stuck when the server comes back, which is usually fine, but in your case there should not be any missing heartbeat errors. So that should go away once I fix the issue with heartbeats (I will test if that's the case).

Again, thanks for adopting the new API already - it's early days so as you see there are some bugs still, but I believe this API is a step in the right direction.

Let me know if I answered your questions and I'll surely let you know as soon as I fix the issue with heartbeats.

**Comment by mdawar** [comment]
@piotrpio yes you did answer all of my questions, thank you very much for the clarifications.

**Comment by mdawar** [comment]
The heartbeat problem was fixed by #1345.
